File: practicas/practica1_rb/Dice.rb
#encoding: utf-8
require_relative "GameCharacter"

module Deepspace
  
  class Dice
    
    def initialize
      @NHANGARSPROB = 0.25
      @NSHIELDSPROB = 0.25
      @NWEAPONSPROB = 0.33
      @FIRSTSHOTPROB = 0.5
      @generator = Random.new
    end

    def initWithNHangars
      if @generator.rand 1.0 < @NHANGARSPROB
        return 0
      else 
        return 1
      end
    end

    def initWithNWeapons
      guess = @generator.rand(1.0)
      if guess < @NWEAPONSPROB
        return 1
      elsif guess < 2 * @NWEAPONSPROB
        return 2
      else
        return 3
      end
    end

    def initWithNShields
      if @generator.rand(1.0) < @NSHIELDSPROB
        return 0
      else
        return 1
      end
    end

    def whoStarts(nPlayers)
      @generator.rand(nPlayers)
    end

    def firstShot
      if @generator.rand(1.0) < @FIRSTSHOTPROB
        return GameCharacter::SPACESTATION
      else
        return GameCharacter::ENEMYSTARSHIP
      end
    end

    def spaceStationMoves(speed)
      @generator.rand 1.0 < speed
    end

  end

end

File: practicas/practica1/src/main/java/deepspace/Dice.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package deepspace;

import java.util.Random;

/**
 *  tiene la responsabilidad de tomar todas las decisiones que dependen del azar en el juego.
 */
class Dice {
    private final float NHANGARSPROB;
    private final float NSHIELDSPROB;
    private final float NWEAPONSPROB;
    private final float FIRSTSHOTPROB;
    private Random generator;
    
    Dice() {
        generator = new Random();
        NHANGARSPROB=0.25f;
        NSHIELDSPROB=0.25f;
        NWEAPONSPROB=0.33f;
        FIRSTSHOTPROB=0.5f;
    }
    
    /**
        Este método determina el número de hangares que recibirá una estación espacial al ser creada
        @return 0 con probabilidad o 1 en otro caso
    */
    public int initWithNHangars() {
        if(generator.nextFloat() < NHANGARSPROB){
            return 0;
        }
        return 1;
    }
    
    /**
     determina el número de armas que recibirá una estación espacial al ser creada
     @return 1, 2 o 3
     */
    public int initWithNWeapons(){
        float guess = generator.nextFloat();
        if(guess < NWEAPONSPROB){
            return 1;
        } else if(guess < 2* NWEAPONSPROB){
            return 2;
        } else {
            return 3;
        }
    }
    
    /**
     * determina el número de potenciadores de escudo que recibirá una estación espacial alser creada
     * @return 0 o 1
     * */
    public int initWithNShields() {
        if(generator.nextFloat() < NSHIELDSPROB){
            return 0;
        }
        
        return 1;
    }
    
    /**
     * determina el jugador (su índice) que iniciará la partida
     * @return numero en rango [0,nplayers-1]
     */
    public int whoStarts(int nPlayers){
        return generator.nextInt(nPlayers);
    }
    
    /**
     * Este método determina quién (de los dos tipos de personajes deljuego) dispara primero en un combate
     * @return GameCharacter.SPACESTATION o GameCharacter.ENEMYSTARSHIP
     * @see GameCharacter
     */
    public GameCharacter firstShot(){
        if(generator.nextFloat()<FIRSTSHOTPROB){
            return GameCharacter.SPACESTATION;
        } else {
            return GameCharacter.ENEMYSTARSHIP;
        }
    }
    
    /**
     * determina si la estaciónespacial se moverá para esquivar un disparo
     * @return un booleano
     */
    public boolean spaceStationMoves(float speed) {
        return (generator.nextFloat() < speed);
    }
}
